geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/partitions/ReceiptTimeScheme.scala

```diff
@@ -0,0 +1,294 @@
+/***********************************************************************
+ * Copyright (c) 2013-2024 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.fs.storage.common.partitions
+
+import org.geotools.filter.visitor.DuplicatingFilterVisitor
+import org.geotools.temporal.`object`.{DefaultInstant, DefaultPeriod, DefaultPosition}
+import org.locationtech.geomesa.fs.storage.api.PartitionScheme.SimplifiedFilter
+import org.locationtech.geomesa.fs.storage.api.{NamedOptions, PartitionScheme, PartitionSchemeFactory}
+import org.locationtech.geomesa.fs.storage.common.partitions.ReceiptTimeScheme.BufferingFilterVisitor
+import org.locationtech.geomesa.utils.geotools.converters.FastConverter
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter._
+import org.opengis.filter.expression.{Expression, Literal, PropertyName}
+import org.opengis.filter.temporal.{After, Before, During, TEquals}
+import org.opengis.temporal.{Instant, Period}
+
+import java.util.Date
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.Duration
+
+/**
+ * Scheme for partitioning based on "receipt time", i.e. when a message is received. Generally this is useful
+ * only for reading existing data that may have been aggregated and stored by an external process.
+ *
+ * @param delegate delegate date time scheme options
+ * @param buffer amount of time to buffer queries by, in order to match a feature date to a receipt time date -
+ *               i.e. the amount of latency in the ingest process
+ */
+case class ReceiptTimeScheme(delegate: DateTimeScheme, buffer: Duration) extends PartitionScheme {
+
+  override val depth: Int = delegate.depth
+
+  override val pattern: String = delegate.pattern
+
+  override def getPartitionName(feature: SimpleFeature): String = delegate.getPartitionName(feature)
+
+  override def getSimplifiedFilters(filter: Filter, partition: Option[String]): Option[Seq[SimplifiedFilter]] = {
+    delegate.getSimplifiedFilters(buffered(filter), partition).map { filters =>
+      // always use the full filter since our dates are not guaranteed to match the partition bounds
+      filters.map(f => f.copy(filter = filter))
+    }
+  }
```

Comment:
Is there a particular reason we're using a delegation pattern here, instead of extending the `DateTimeScheme` class? I think we can just inherit `depth`, `pattern`, and `getPartitionName` directly from the `DateTimeScheme` class instead of having to override them.

Reply:
hmm yeah, no particular reason - i was thinking of this as a separate thing from the date time scheme, rather than an extension, but in practice it would be equivalent. I guess i mainly didn't want to deal with the constructor args required for the date time scheme, i.e. it would need to change to `class ReceiptTimeScheme(buffer: Duration, formatter: DateTimeFormatter, stepUnit: ChronoUnit, step: Int, dtg: String, dtgIndex: Int) extends DateTimeScheme(formatter, stepUnit, step, dtg, dtgIndex)`